### PR TITLE
[SPARK-14647][SQL] Group SQLContext/HiveContext state into SharedState

### PR DESCRIPTION
**Pull request by andrewor14**
## What changes were proposed in this pull request?

This patch adds a `PersistentState` that groups state shared across multiple `SQLContext`s. This is analogous to the `SessionState` added in SPARK-13526 that groups session-specific state. This cleanup makes the constructors of the contexts simpler and ultimately allows us to remove `HiveContext` in the near future.
## How was this patch tested?

Just refactoring.

### CODE REVIEW

**Comment by rxin** [comment]
We need to find a different name -- the states are not necessarily persistent (e.g. cache manager).

**Comment by yhuai** [comment]
```
2016-04-14 18:24:27.954 - stderr> Exception in thread "main" java.lang.NullPointerException
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.SQLContext.sparkContext(SQLContext.scala:100)
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.SQLContext.<init>(SQLContext.scala:81)
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:83)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:88)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLEnv$.init(SparkSQLEnv.scala:59)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.<init>(SparkSQLCLIDriver.scala:281)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.main(SparkSQLCLIDriver.scala:138)
2016-04-14 18:24:27.961 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.main(SparkSQLCLIDriver.scala)
2016-04-14 18:24:27.961 - stderr> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2016-04-14 18:24:27.961 - stderr> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2016-04-14 18:24:27.962 - stderr> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2016-04-14 18:24:27.962 - stderr> 	at java.lang.reflect.Method.invoke(Method.java:497)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:726)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:183)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:208)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:122)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

**Comment by andrewor14** [comment]
Hm, `SharedState`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala [hunk 1]
```diff
@@ -114,25 +94,25 @@ class HiveContext private[hive](
 
   logDebug("create HiveContext")
 
+  @transient
+  protected[sql] override val persistentState: PersistentState = hivePersistentState
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala]: This line caused the NPE. We can just remove it, right?

**Comment by yhuai** [comment]
`SharedState` sounds better.

**Comment by andrewor14** [comment]
`GlobalState`

**Comment by rxin** [comment]
either works --

**Comment by andrewor14** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala [hunk 2]
```diff
@@ -549,6 +505,22 @@ private[hive] class TestHiveFunctionRegistry extends SimpleFunctionRegistry {
...
 
+
+private[hive] class TestHiveSharedState(
+    sc: SparkContext,
+    warehousePath: File,
+    scratchDirPath: File,
+    metastoreTemporaryConf: Map[String, String])
+  extends HiveSharedState(sc) {
+
+  override val metadataHive: HiveClient = {
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala]: I figured out the issue with the previous test failure. Although we are trying to override it at here, but inside HiveSharedState, we still create a metadataHive which uses the derby `metastore_db`.

**Comment by andrewor14** [comment]
retest this please

**Comment by yhuai** [comment]
looks like a legitimate failure?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSharedState.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   * A Hive client used for execution.
+   */
+  val executionHive: HiveClientImpl = {
+    HiveContext.newClientForExecution(sparkContext.conf, sparkContext.hadoopConfiguration)
+  }
+
+  /**
+   * A Hive client used to interact with the metastore.
+   */
+  lazy val metadataHive: HiveClient = {
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSharedState.scala]: Let's explain why we need to make it a lazy val at here.

**Pull request closed without merging**